interface/TausProducer.h

```diff
@@ -0,0 +1,151 @@
+#ifndef TAUS_PRODUCER
+#define TAUS_PRODUCER
+
+#include <cp3_llbb/Framework/interface/LeptonsProducer.h>
+#include <cp3_llbb/Framework/interface/ScaleFactors.h>
+
+#include <DataFormats/VertexReco/interface/Vertex.h>
+#include <DataFormats/PatCandidates/interface/Tau.h>
+#include <DataFormats/PatCandidates/interface/TauPFSpecific.h>
+
+#include <utility>
+#include <Math/VectorUtil.h>
+
+typedef std::unordered_map<std::string, float> tauDiscriminatorMap;
+typedef ROOT::Math::LorentzVector<ROOT::Math::PtEtaPhiE4D<float> > LorentzVector;
+
+enum class GenMatch { Electron = 1, Muon = 2, TauElectron = 3,  TauMuon = 4, Tau = 5, NoMatch = 6 };
+
+namespace tools {
+  template<typename Type>
+    std::set<Type> union_sets(std::initializer_list<std::set<Type>> sets)
+    {
+      std::set<Type> result;
+      for(const auto& set : sets)
+	result.insert(set.begin(), set.end());
+      return result;
+    }
+}
+
+using MatchResult = std::pair<GenMatch, const reco::GenParticle*>;
+
+
+class TausProducer: public LeptonsProducer<pat::Tau>, public ScaleFactors {
+    public:
+        TausProducer(const std::string& name, const ROOT::TreeGroup& tree, const edm::ParameterSet& config):
+            LeptonsProducer(name, tree, config), ScaleFactors(const_cast<ROOT::TreeGroup&>(tree))
+        {
+            ScaleFactors::create_branches(config);
+        }
+
+        virtual ~TausProducer() {}
+
+        virtual void doConsumes(const edm::ParameterSet& config, edm::ConsumesCollector&& collector) override {
+            LeptonsProducer::doConsumes(config, std::forward<edm::ConsumesCollector>(collector));
+
+            m_vertices_token = collector.consumes<std::vector<reco::Vertex>>(config.getUntrackedParameter<edm::InputTag>("vertices", edm::InputTag("offlineSlimmedPrimaryVertices")));
+	    m_pruned_token = collector.consumes<std::vector<reco::GenParticle>>(config.getUntrackedParameter<edm::InputTag>("pruned_gen_particles", edm::InputTag("prunedGenParticles")));
+	}
+
+        virtual void produce(edm::Event& event, const edm::EventSetup& eventSetup) override;
+
+	
+    private:
+        // Tokens
+        edm::EDGetTokenT<std::vector<reco::Vertex>> m_vertices_token;
+	edm::EDGetTokenT<std::vector<reco::GenParticle>> m_pruned_token;
+    private:
+ 
+	MatchResult LeptonGenMatch(const LorentzVector& p4, const std::vector<reco::GenParticle>& genParticles);
+
+	inline void FindFinalStateDaughters(const reco::GenParticle& particle, std::vector<const reco::GenParticle*>& daughters,
+					    const std::set<int>& pdg_to_exclude = {})
+	{
+	  if(!particle.daughterRefVector().size()) {
```

Comment:
detail: `if( particle.daughterRefVector().empty() )` (more efficient and more readable)